### ISSUE TITLE
[rush] proposal : add a "type" field on ```rush change```

### ISSUE THREAD

**Issue by WanderWang** [microsoft/rushstack#2606]
### Our workflow currently

out team use use rush with Git Flow  : 
* use GitLab ( include GitLab , GitLab-CI and GitLab-Runner )
* master branch is protected branch , only MergeRequest ( such as PullRequest in GitHub ) can be merged into master
* developer commit their code in features branch , and create a MergeRequest , then nofity the owner of the monorepo to merge the MR.
* before create a MR , developer must execute ```rush change``` to author the changelog
* when MR created , GitLab pipeline will execute some check ,only MR that has passed the check can be merged , the check inludes : 
    * eslint
    * prettier
    * rush change -v (  check if ```rush change``` is executed )
* the owner of monorepo merge the MR 
* the owner of monorepo execute ```rush publish -a ``` on master branch to generate fully changelog .

above the workflow works very well , but our team want to enhance it on changelog generate .

#### Lint and Hooks

1. because of changelog is authored by developers , it should be ' lint' , ( for example ,changelog must more than 10 charater )
2. lint should be customed by the owner of monorepo ,here is a example 
```javascript
//common/scripts/change-log-hooks

function format(text){
    return text.trim();
}

module.exports = ( changelog ) => {
    changelog.comment = format(changelog.comment);
    if (changelog.comment.length <= 10) {
        throw new Error(",changelog must more than 10 charater ")
    }
    return changelog
}
```

#### ChangeLog should have a "type" field 

 I want to create a changelog file for my customer automantic generate via changelog.json ,but I don't want to add engineering / refactor into it .

here is my tentative idea 
1. when developer execute ```rush change``` , developer must select a type ( feature / bugfix eg..) , and changelog.json should have a type field .
2. "type" should be a choice not a string input 
3. maybe I can customize type in a configfile.

this feature can be integrated into rush or customized with Lint and Hooks , such as : 

```javascript
//common/scripts/change-log-hooks

function extractType(text){
    return text.split(":")[0]
}

module.exports = ( changelog ) => {
    changelog.type= extractType(changelog.comment);
    return changelog;
}
```

**Comment by octogonz** [comment]
👉  The original issue proposed two separate features.  Although they have some interaction, the design/implementation can be done independently, so I've split them into two issues:  #2606 tracks the configurable change types.  And #2609 tracks the script hook.

**Comment by octogonz** [comment]
It's pretty straightforward to add a `"type"` field to the JSON "change files" (under **common/changes**).

The more interesting design question is how the options would be specified.  For example, [version-policies.json](https://rushjs.io/pages/configs/version-policies_json/) already allows you to customize `rush change` behavior via settings  such as:

- `"exemptFromRushChange"`
- `"lockedMajor"`
- `"mainProject"`

But those settings are only accessible if you use a "version policy".

This spring I have been putting together some plans to (1) decouple `rush change` so that it is usable without publishing NPM packages and (2) redesign the config files, converting most of the CLI parameters for `"rush version"` and `"rush publish"` into config file settings.  It will be implemented as a series of small steps, that are opt-in via **common/config/experiments.json**.

So rather than adding more options to **version-policies.json**, maybe we could use this #2606 to set up a new file that is just for versioning and `rush change`.  I am on vacation right now, but I will be back next week, and then I can write up a suggested config file design.

**Comment by iclanton** [comment]
I agree with @octogonz - it makes sense for this to happen with the revamping of publishing.

**Comment by renoirb** [comment]
That is great!

[Question]: Now I wonder how publishing works in sich a context. How do you handle publishing to private NPM and git tag with version bump during publish?
In your team, do you enforce any merge to be approved and do you require to publish N changed packages that are publishable on merge?
